File: src/runtime/jsonrpc/connection.ts
import type { Duplex } from "node:stream"
import { encodeMessage, FrameDecoder } from "./framing.ts"
import {
  isFailure,
  isResponse,
  JsonRpcErrorCode,
  RpcError,
  type JsonRpcErrorObject,
  type JsonRpcId,
  type JsonRpcMessage,
  type JsonRpcRequest,
  type JsonRpcResponse,
} from "./messages.ts"

// Handles one inbound request or notification. Return a value to answer a
// request; throw `RpcError` (or any error) to fail it. The return value is
// ignored for notifications. The same handler serves both directions, so a
// peer can act as client and server at once (bidirectional — plan §4).
export type RpcHandler = (
  method: string,
  params: unknown,
) => unknown | Promise<unknown>

type Pending = {
  resolve: (value: unknown) => void
  reject: (reason: unknown) => void
}

// A full-duplex JSON-RPC 2.0 endpoint over any Node duplex stream (a UDS
// socket, a child process stdio pair, or an in-memory pipe in tests).
export class JsonRpcConnection {
  private readonly decoder = new FrameDecoder()
  private readonly pending = new Map<JsonRpcId, Pending>()
  private nextId = 1
  private handler: RpcHandler | undefined
  private closed = false

  constructor(
    private readonly stream: Duplex,
    handler?: RpcHandler,
  ) {
    this.handler = handler
    stream.on("data", (chunk: Buffer) => this.onData(chunk))
    stream.on("error", (error) => this.onClose(error))
    stream.on("close", () => this.onClose(new Error("Connection closed.")))
  }

  onRequest(handler: RpcHandler): void {
    this.handler = handler
  }

  request<Result = unknown>(method: string, params?: unknown): Promise<Result> {
    return new Promise<Result>((resolve, reject) => {
      if (this.closed) {
        reject(
          new RpcError(JsonRpcErrorCode.InternalError, "Connection is closed."),
        )
        return
      }

      const id = this.nextId++
      this.pending.set(id, {
        resolve: resolve as (value: unknown) => void,
        reject,
      })
      this.send({ jsonrpc: "2.0", id, method, params })
    })
  }

  notify(method: string, params?: unknown): void {
    if (this.closed) return
    this.send({ jsonrpc: "2.0", method, params })
  }

  close(): void {
    this.onClose(new Error("Connection closed by caller."))
    this.stream.end()
  }

  private send(message: JsonRpcMessage): void {
    this.stream.write(encodeMessage(message))
  }

  private onData(chunk: Buffer): void {
    this.decoder.append(chunk)
    for (const message of this.decoder.drain()) {
      this.dispatch(message)
    }
  }

  private dispatch(message: JsonRpcMessage): void {
    if (isResponse(message)) {
      this.resolveResponse(message)
      return
    }

    if ("id" in message && message.id !== undefined) {
      void this.handleRequest(message)
    } else {
      void this.handleNotification(message.method, message.params)
    }
  }

  private resolveResponse(response: JsonRpcResponse): void {
    const pending = this.pending.get(response.id)
    if (!pending) return

    this.pending.delete(response.id)
    if (isFailure(response)) {
      const { code, message, data } = response.error
      pending.reject(new RpcError(code, message, data))
    } else {
      pending.resolve(response.result)
    }
  }

  private async handleRequest(request: JsonRpcRequest): Promise<void> {
    try {
      const result = await this.invoke(request.method, request.params)
      this.send({ jsonrpc: "2.0", id: request.id, result: result ?? null })
    } catch (error) {
      this.send({
        jsonrpc: "2.0",
        id: request.id,
        error: toErrorObject(error),
      })
    }
  }

  private async handleNotification(
    method: string,
    params: unknown,
  ): Promise<void> {
    try {
      await this.invoke(method, params)
    } catch {
      // Notifications have no response channel; swallow handler errors.
    }
  }

  private invoke(method: string, params: unknown): unknown | Promise<unknown> {
    if (!this.handler) {
      throw new RpcError(
        JsonRpcErrorCode.MethodNotFound,
        `No handler for "${method}".`,
      )
    }
    return this.handler(method, params)
  }

  private onClose(reason: unknown): void {
    if (this.closed) return
    this.closed = true
    for (const pending of this.pending.values()) {
      pending.reject(reason)
    }
    this.pending.clear()
  }
}

const toErrorObject = (error: unknown): JsonRpcErrorObject => {
  if (error instanceof RpcError) return error.toErrorObject()
  const message = error instanceof Error ? error.message : String(error)
  return { code: JsonRpcErrorCode.InternalError, message }
}

File: tui/internal/jsonrpc/connection.go
package jsonrpc

import (
	"bufio"
	"context"
	"encoding/json"
	"io"
	"strconv"
	"sync"
	"sync/atomic"
)

// Handler handles one inbound request or notification. Return a value to answer
// a request; return an error (an *Error to set a specific code/data, or any
// error for InternalError) to fail it. The return value is ignored for
// notifications. The same handler serves both directions, so a peer can act as
// client and server at once (bidirectional — plan §4).
type Handler func(method string, params json.RawMessage) (any, error)

// Conn is a full-duplex JSON-RPC 2.0 endpoint over any io.ReadWriteCloser (a UDS
// socket, a child process stdio pair, or an in-memory pipe in tests).
type Conn struct {
	rw      io.ReadWriteCloser
	reader  *bufio.Reader
	handler Handler

	writeMu sync.Mutex

	mu      sync.Mutex
	nextID  int64
	pending map[int64]chan *Message
	closed  bool
	closeFn sync.Once

	// Notifications are dispatched by a single worker so they run in arrival
	// order — streaming events (e.g. AI message chunks) must not be reordered.
	notifications chan *Message
	done          chan struct{}
}

// NewConn starts a connection and its read loop. handler may be nil if this peer
// only makes outbound calls.
func NewConn(rw io.ReadWriteCloser, handler Handler) *Conn {
	c := &Conn{
		rw:            rw,
		reader:        bufio.NewReader(rw),
		handler:       handler,
		pending:       make(map[int64]chan *Message),
		notifications: make(chan *Message, 256),
		done:          make(chan struct{}),
	}
	go c.readLoop()
	go c.notificationLoop()
	return c
}

// Request sends a request and blocks until the response arrives or ctx is done.
func (c *Conn) Request(ctx context.Context, method string, params any) (json.RawMessage, error) {
	raw, err := marshalParams(params)
	if err != nil {
		return nil, err
	}

	id := atomic.AddInt64(&c.nextID, 1)
	ch := make(chan *Message, 1)

	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil, &Error{Code: InternalError, Message: "connection is closed"}
	}
	c.pending[id] = ch
	c.mu.Unlock()

	idRaw := json.RawMessage(strconv.FormatInt(id, 10))
	if err := c.write(&Message{JSONRPC: "2.0", ID: &idRaw, Method: method, Params: raw}); err != nil {
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, err
	}

	select {
	case <-ctx.Done():
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, ctx.Err()
	case resp := <-ch:
		if resp.Error != nil {
			return nil, resp.Error
		}
		return resp.Result, nil
	}
}

// Notify sends a fire-and-forget notification (no response expected).
func (c *Conn) Notify(method string, params any) error {
	raw, err := marshalParams(params)
	if err != nil {
		return err
	}
	return c.write(&Message{JSONRPC: "2.0", Method: method, Params: raw})
}

// Close shuts the connection and fails every in-flight request.
func (c *Conn) Close() error {
	c.shutdown(&Error{Code: InternalError, Message: "connection closed by caller"})
	return c.rw.Close()
}

func (c *Conn) write(msg *Message) error {
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	return writeMessage(c.rw, msg)
}

func (c *Conn) readLoop() {
	for {
		msg, err := readMessage(c.reader)
		if err != nil {
			c.shutdown(&Error{Code: InternalError, Message: err.Error()})
			return
		}
		c.dispatch(msg)
	}
}

func (c *Conn) dispatch(msg *Message) {
	if msg.isResponse() {
		c.resolveResponse(msg)
		return
	}
	if msg.ID != nil {
		go c.handleRequest(msg)
	} else {
		// Queue for the ordered worker; drop if the connection is closing.
		select {
		case c.notifications <- msg:
		case <-c.done:
		}
	}
}

// notificationLoop processes notifications one at a time, in arrival order.
func (c *Conn) notificationLoop() {
	for {
		select {
		case msg := <-c.notifications:
			c.handleNotification(msg)
		case <-c.done:
			return
		}
	}
}

func (c *Conn) resolveResponse(msg *Message) {
	id, err := strconv.ParseInt(string(*msg.ID), 10, 64)
	if err != nil {
		return
	}
	c.mu.Lock()
	ch := c.pending[id]
	delete(c.pending, id)
	c.mu.Unlock()
	if ch != nil {
		ch <- msg
	}
}

func (c *Conn) handleRequest(msg *Message) {
	resp := &Message{JSONRPC: "2.0", ID: msg.ID}
	result, err := c.invoke(msg)
	if err != nil {
		resp.Error = toError(err)
	} else {
		raw, marshalErr := json.Marshal(result)
		if marshalErr != nil {
			resp.Error = &Error{Code: InternalError, Message: marshalErr.Error()}
		} else {
			resp.Result = raw
		}
	}
	_ = c.write(resp)
}

func (c *Conn) handleNotification(msg *Message) {
	// Notifications have no response channel; ignore handler errors.
	_, _ = c.invoke(msg)
}

func (c *Conn) invoke(msg *Message) (any, error) {
	if c.handler == nil {
		return nil, &Error{Code: MethodNotFound, Message: "no handler for " + msg.Method}
	}
	return c.handler(msg.Method, msg.Params)
}

func (c *Conn) shutdown(reason *Error) {
	c.closeFn.Do(func() {
		close(c.done)
		c.mu.Lock()
		c.closed = true
		pending := c.pending
		c.pending = make(map[int64]chan *Message)
		c.mu.Unlock()
		for _, ch := range pending {
			ch <- &Message{JSONRPC: "2.0", Error: reason}
		}
	})
}

func marshalParams(params any) (json.RawMessage, error) {
	if params == nil {
		return nil, nil
	}
	return json.Marshal(params)
}

// toError maps a handler error to a JSON-RPC error object. An *Error passes
// through unchanged; anything else becomes InternalError.
func toError(err error) *Error {
	if rpcErr, ok := err.(*Error); ok {
		return rpcErr
	}
	return &Error{Code: InternalError, Message: err.Error()}
}
